internal/rbd/rbd_util.go

```diff
@@ -1441,6 +1441,43 @@ func (rv *rbdVolume) isThickProvisioned() (bool, error) {
 	return thick, nil
 }
 
+// RepairThickProvision writes zero bytes to the volume so that it will be
+// completely allocated. In case the volume is already marked as
+// thick-provisioned, nothing will be done.
+func (rv *rbdVolume) RepairThickProvision() error {
+	// if the image has the thick-provisioned metadata, it has been fully
+	// allocated
+	done, err := rv.isThickProvisioned()
+	if err != nil {
+		return fmt.Errorf("failed to repair thick-provisioning of %q: %w", rv, err)
+	} else if done {
+		return nil
+	}
+
+	// in case there are watchers, assume allocating is still happening in
```

Comment:
nit: s/allocating/allocation
btw, If there are no watchers on the image, how the allocation can still happen @nixpanic ?  @idryomov do we expect such conditions in any stage ?

Reply:
_allocating_ - the action is in progress?

The leader provisioner Pod could have lost its lease for some reason, and the previously leading Pod may still be working on it.

Reply:
@nixpanic I think the point here is mainly on this: `how we (could land)/landed on more than one watcher` and if we land to a situation how would we identify what caused it, to make  sure or validate whether  is ongoing provisioning is  a genuine attempt happening or a left over from a stale request? If disturbances like leader reelection causes it ( I doubt that could leave such state) , If its from a stale/unmapped request for the CO, the operation may succeed in background , but we are going to retry the same in next request with new leader or whoever called, its going to be an overlap or unwanted second attempt. Suppose the first request caused the watcher to be there, but its no longer functioning , is this watcher get removed automatically after some time ? if not, we cant operate on this volume any more, that would also leave a stale volume in the backend as we are not doing any cleanup here.  I am just trying to understand the workflow in a bad situation , however I think its good to run through so that we are better prepared.

Reply:
Yes, the watcher gets removed automatically. See the log snippet I added for https://github.com/ceph/ceph-csi/pull/2101#discussion_r640422026 where the 2nd provisioner restarts the CreateVolume request and re-allocates the RBD image. Initially the watch is still present and allocating aborts, but the retries for CreateVolume continue shortly after.

Reply:
An inactive watch should be cleaned up after 30 seconds.